# Задание 1. Реализовать метод по примеру для расчета дисперсий к следующей задаче.
Решить задачу, используя методы Anova SciPy.
Провести дисперсионный анализ для определения того, есть ли различия среднего роста среди взрослых футболистов, хоккеистов и штангистов. Даны значения роста в трех группах случайно выбранных спортсменов: Футболисты: 173, 175, 180, 178, 177, 185, 183, 182. Хоккеисты: 177, 179, 180, 188, 177, 172, 171, 184, 180. Штангисты: 172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170.

In [355]:
import numpy as np

In [356]:
football = [173,175, 180, 178, 177, 185, 183, 182]
hockey = [177, 179, 180, 188, 177, 172, 171, 184, 180]
weightlifters=[172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170]

In [357]:
n1 = len(football)
n2 = len(hockey)
n3 = len(weightlifters)
n = n1 + n2 + n3
n

28

Всего три группы:

In [84]:
k = 3

Проведем однофакторный дисперсионный анализ.Посчитаем среднее значение каждого массива:

In [85]:
football_mean = np.mean(football)
football_mean

179.125

In [86]:
hockey_mean = np.mean(hockey)
hockey_mean

178.66666666666666

In [87]:
weightlifters_mean = np.mean(weightlifters)
weightlifters_mean

172.72727272727272

Видно, что среднее значение роста в трех группах разнятся. Установим, что это отличие статистически значимо. Для этого сначала соберем все значения роста в один массив:

In [88]:
height_all = np.concatenate([football,hockey,weightlifters])
height_all

array([173, 175, 180, 178, 177, 185, 183, 182, 177, 179, 180, 188, 177,
       172, 171, 184, 180, 172, 173, 169, 177, 166, 180, 178, 177, 172,
       166, 170])

Найдем среднее значение:

In [89]:
height_mean = np.mean(height_all)
height_mean

176.46428571428572

Найдем $S^2$ — сумму квадратов отклонений наблюдений от общего среднего:

In [90]:
s2 = np.sum((height_all - height_mean)**2)
s2

830.9642857142854

Найдем $S^2_F$ - сумму квадратов отклонений средних групповых значений от общего среднего:

In [91]:
s2_f = ((football_mean - height_mean)**2) * n1 + ((hockey_mean - height_mean)**2) * n2 + ((weightlifters_mean - height_mean)**2) * n3
s2_f

253.9074675324678

Найдем $S^2_{ост}$ — остаточную сумму квадратов отклонений:

In [92]:
s2_residual = np.sum((football - football_mean)**2) + np.sum((hockey - hockey_mean)**2) + np.sum((weightlifters - weightlifters_mean)**2)
s2_residual

577.0568181818182

Удостоверимся, что соблюдается равенство $S^2 = S_F^2 + S_{ост}^2$:

In [93]:
print(s2)
print(s2_f + s2_residual)

830.9642857142854
830.964285714286


Найдем общую дисперсию:

In [94]:
sigma2_general = s2 / (n - 1)
sigma2_general

30.776455026455015

Найдем факторную дисперсию:

In [95]:
sigma2_f = s2_f / (k - 1)
sigma2_f

126.9537337662339

Найдем остаточную дисперсию:

In [96]:
sigma2_residual = s2_residual / (n - k)
sigma2_residual

23.08227272727273

Вычислим $F_H$:

In [97]:
F_h = sigma2_f / sigma2_residual
F_h

5.500053450812598

Найдем значение $F_{крит}$ в таблице критических точек распределения Фишера-Снедекора для уровня значимости $\alpha = 0.05$ и двух степеней свободы:

$df_{межд} = k - 1 = 3 - 1 = 2$ и $df_{внутр} = n - k = 28 - 3 = 25$.

Для данных значений $F_{крит} = 3,38$. Так как $F_H$>$ F_{крит}$, гипотеза Н0  не принимается, различие роста в трех группах статистически значимо.

# Задание2. Проанализируйте набор данных «Авто-мили», чтобы выяснить, насколько переменные похожи друг на друга.
Загрузите набор данных auto-mpg из ~ / наборов данных / auto-mpg / Проверить нулевую гипотезу, отклонять или не отклонять нулевую гипотезу.

Гипотезы:
Independent variable: origin

Dependent variable: mpg

$ H_o: $ Существенных различий в милях среди транспортных средств различного происхождения не существует.

$ H_1: $ Существенная разница существует в милях среди транспортных средств различного происхождения.

Independent variable: origin

Independent variable: horsepower

Dependent variable: mpg

$ H_o: $ Существенных различий в милях среди транспортных средств различного происхождения не существует.

$ H_1: $ Существенная разница существует в милях среди транспортных средств различного происхождения.

In [358]:
import pandas as pd

In [359]:
df = pd.read_csv('auto-mpg.csv')
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

#### Для начала проверим когда зависимая переменная только origin.

Сгруппируем по значениям, которые присутствуют в этой переменной и выведем соответсвущее количество mpg.

In [360]:
origin = df.groupby('origin')['mpg'].count()
origin

origin
1    249
2     70
3     79
Name: mpg, dtype: int64

In [361]:
origin1 = df.groupby('origin').mpg.apply(list)[1]
origin2 = df.groupby('origin').mpg.apply(list)[2]
origin3 = df.groupby('origin').mpg.apply(list)[3]

In [362]:
n1 = origin[1]
n2 = origin[2]
n3 = origin[3]
n = n1+n2+n3
n

398

Всего три группы:

In [363]:
k = 3

Проведем однофакторный дисперсионный анализ.Посчитаем среднее значение каждого массива:

In [364]:
origin1_mean = np.mean(origin1)
origin1_mean

20.083534136546184

In [365]:
origin2_mean = np.mean(origin2)
origin2_mean

27.89142857142857

In [366]:
origin3_mean = np.mean(origin3)
origin3_mean

30.450632911392404

Видно, что среднее значение роста в трех группах разнятся. Установим, что это отличие статистически значимо. Для этого сначала соберем все значения роста в один массив:

In [367]:
origin_all = np.concatenate([origin1,origin2,origin3])
origin_all

array([18. , 15. , 18. , 16. , 17. , 15. , 14. , 14. , 14. , 15. , 15. ,
       14. , 15. , 14. , 22. , 18. , 21. , 21. , 10. , 10. , 11. ,  9. ,
       28. , 25. , 19. , 16. , 17. , 19. , 18. , 14. , 14. , 14. , 14. ,
       12. , 13. , 13. , 18. , 22. , 19. , 18. , 23. , 26. , 25. , 20. ,
       21. , 13. , 14. , 15. , 14. , 17. , 11. , 13. , 12. , 13. , 15. ,
       13. , 13. , 14. , 22. , 28. , 13. , 14. , 13. , 14. , 15. , 12. ,
       13. , 13. , 14. , 13. , 12. , 13. , 18. , 16. , 18. , 18. , 23. ,
       11. , 12. , 13. , 12. , 18. , 21. , 19. , 21. , 15. , 16. , 15. ,
       11. , 20. , 21. , 19. , 15. , 26. , 25. , 16. , 16. , 18. , 16. ,
       13. , 14. , 14. , 14. , 28. , 19. , 18. , 15. , 15. , 16. , 15. ,
       16. , 14. , 17. , 16. , 15. , 18. , 21. , 20. , 13. , 23. , 20. ,
       23. , 18. , 19. , 25. , 26. , 17.5, 16. , 15.5, 14.5, 22. , 22. ,
       24. , 22.5, 29. , 24.5, 20. , 18. , 18.5, 17.5, 26.5, 13. , 16.5,
       13. , 13. , 13. , 30. , 25.5, 17.5, 17. , 15

Найдем среднее значение:

In [368]:
origin_mean = np.mean(origin_all)
origin_mean

23.514572864321607

Найдем $S^2$  — сумму квадратов отклонений наблюдений от общего среднего:

In [369]:
s2 = np.sum((origin_all - origin_mean)**2)
s2

24252.575477386934

Найдем $S^2_{F}$ - сумму квадратов отклонений средних групповых значений от общего среднего:

In [371]:
s2_f = ((origin1_mean - origin_mean)**2) * n1 + ((origin2_mean - origin_mean)**2) * n2 + ((origin3_mean - origin_mean)**2) * n3
s2_f

8072.820661929806

Найдем $S^2_{ост}$ — остаточную сумму квадратов отклонений:

In [372]:
s2_residual = np.sum((origin1 - origin1_mean)**2) + np.sum((origin2 - origin2_mean)**2) + np.sum((origin3 - origin3_mean)**2)
s2_residual

16179.754815457129

Удостоверимся, что соблюдается равенство $S^2 = S_F^2 + S_{ост}^2$:

In [373]:
print(s2)
print(s2_f + s2_residual)

24252.575477386934
24252.575477386934


Найдем общую дисперсию:

In [374]:
sigma2_general = s2 / (n - 1)
sigma2_general

61.08961077427439

Найдем факторную дисперсию:

In [375]:
sigma2_f = s2_f / (k - 1)
sigma2_f

4036.410330964903

Найдем остаточную дисперсию:

In [376]:
sigma2_residual = s2_residual / (n - k)
sigma2_residual

40.961404596094

Вычислим $F_H$:

In [377]:
F_h = sigma2_f / sigma2_residual
F_h

98.54179491075868

Найдем значение $F_{крит}$ в таблице критических точек распределения Фишера-Снедекора для уровня значимости $\alpha = 0.05$ и двух степеней свободы:

$df_{межд} = k - 1 = 3 - 1 = 2$ и $df_{внутр} = n - k = 398 - 3 = 395$.

Для данных значений $F_{крит} = 2,99$. Так как $F_H$>$ F_{крит}$, гипотеза Н0 отвергается, различие роста в трех группах статистически значимо.

In [378]:
##нашла такой способ,решила проверить значения
import scipy.stats as stats

print (stats.f_oneway(origin1,origin2,origin3))

F_onewayResult(statistic=98.54179491075871, pvalue=1.9154864184128e-35)


#### Теперь проверим когда зависимая переменная не только origin, но и horsepower.

In [379]:
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [380]:
df.drop(['cylinders','displacement','weight','acceleration','model year','car name'],axis = 1)

mpg horsepower  origin
0    18.0        130       1
1    15.0        165       1
2    18.0        150       1
3    16.0        150       1
4    17.0        140       1
..    ...        ...     ...
393  27.0         86       1
394  44.0         52       2
395  32.0         84       1
396  28.0         79       1
397  31.0         82       1

[398 rows x 3 columns]

Two-way ANOVA using Statsmodels

In [382]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
import matplotlib.pyplot as plt

In [385]:
# ANOVA
formula = 'mpg~C(origin)+C(horsepower)+C(origin):C(horsepower)'
model = ols(formula, df).fit()
aov_table = anova_lm(model, typ=2)

print(aov_table)

                               sum_sq     df           F        PR(>F)
C(origin)                 3185.081821    2.0  123.199009  9.557802e-24
C(horsepower)            30403.388701   93.0   25.290398  2.792820e-44
C(origin):C(horsepower)  12068.401881  186.0    5.019419  3.542619e-21
Residual                  3412.615119  264.0         NaN           NaN


Нас интересует зависимость mpg от origin -> $F_H$ = 123.19.

Для данных значений $F_{крит} = 2,68$. Так как $F_H$>$ F_{крит}$, гипотеза Н0 не принимается, различие роста в трех группах статистически значимо.

# Задание 3. Определить , какие выборки и какой критерий использовать.

Изучить, оказывает ли влияение новая недавно разработанная система электронной регистрации на прием к врачу на среднее время, проведенное пациентом в ожидании приема. Пусть есть две выборки со средним временем ожидания приема за последний месяц в двух разных клиниках; в одной из них есть система электронной регистрации, а в другой – нет.


#### Independent variable: клиника с разным способом регистрации
#### Dependent variable: среднее время ожидание приема
#### Критерий Фишера

Для тестирования диеты набираются 15 человек. Их вес замеряется, а затем они подвергаются диете в течении некоторого времени, и их вес снова замеряется. С помощью статистических методов необходимо проверить, работает ли диета.

#### Independent variable: время
#### Dependent variable: вес испытуемых
#### Критерий Фишера

Влияние витамина C на рост зубов морских свинок. В эксперименте изучалось влияние витамина C на рост одонтобластов морских свинок. Рассматривалось три дозы витамина (0.5, 1 и 2 мг) и два способа его приёма (апельсиновый сок и аскорбиновая кислота). В каждой из 6 групп по уровням двух факторов были произведены измерения для 10 морских свинок.

#### Independent variable: доза и способы приема
#### Dependent variable: рост
#### Критерий Фишера

Рост певцов хора В 1979 году было опрошено 235 членов нью-йоркской ассоциации хорового пения, для каждого из них известен рост и регистр голоса. Меняется ли средний рост вместе с регистром?

#### Independent variable: регистр
#### Dependent variable: рост
#### Критерий Фишера